src/Features/Core/Portable/EditAndContinue/Remote/RemoteEditAndContinueServiceProxy.cs

```diff
@@ -0,0 +1,462 @@
+            IDiagnosticAnalyzerService diagnosticService,
+            ActiveStatementProvider activeStatementProvider,
+            IDebuggeeModuleMetadataProvider debuggeeModuleMetadataProvider,
+            CancellationToken cancellationToken)
+        {
+            IDisposable result;
+            ImmutableArray<DocumentId> documentsToReanalyze;
+
+            var client = await RemoteHostClient.TryGetClientAsync(_workspace, cancellationToken).ConfigureAwait(false);
+            if (client == null)
+            {
+                GetLocalService().StartEditSession(activeStatementProvider, debuggeeModuleMetadataProvider, out documentsToReanalyze);
+                result = LocalConnection.Instance;
+            }
+            else
+            {
+                // need to keep the providers alive until the edit session ends:
+                var connection = await client.CreateConnectionAsync(
+                    WellKnownServiceHubService.RemoteEditAndContinueService,
+                    callbackTarget: new StartEditSessionCallback(activeStatementProvider, debuggeeModuleMetadataProvider),
+                    cancellationToken).ConfigureAwait(false);
+
+                documentsToReanalyze = await connection.RunRemoteAsync<ImmutableArray<DocumentId>>(
+                    nameof(IRemoteEditAndContinueService.StartEditSessionAsync),
+                    solution: null,
+                    Array.Empty<object>(),
+                    cancellationToken).ConfigureAwait(false);
+
+                result = connection;
+            }
+
+            // clear all reported run mode diagnostics:
+            diagnosticService.Reanalyze(_workspace, documentIds: documentsToReanalyze);
+
+            return result;
+        }
+
+        private sealed class LocalConnection : IDisposable
+        {
+            public static readonly LocalConnection Instance = new LocalConnection();
+
+            public void Dispose()
+            {
+            }
+        }
+
+        private sealed class StartEditSessionCallback : IRemoteEditAndContinueService.IStartEditSessionCallback
+        {
+            private readonly ActiveStatementProvider _activeStatementProvider;
+            private readonly IDebuggeeModuleMetadataProvider _managedModuleInfoProvider;
+
+            public StartEditSessionCallback(ActiveStatementProvider activeStatementProvider, IDebuggeeModuleMetadataProvider debuggeeModuleMetadataProvider)
+            {
+                _activeStatementProvider = activeStatementProvider;
+                _managedModuleInfoProvider = debuggeeModuleMetadataProvider;
+            }
+
+            /// <summary>
+            /// Remote API.
+            /// </summary>
+            public async Task<ImmutableArray<ActiveStatementDebugInfo>> GetActiveStatementsAsync(CancellationToken cancellationToken)
+            {
+                try
+                {
+                    return await _activeStatementProvider(cancellationToken).ConfigureAwait(false);
```

Comment:
this is the part that goes away later with Sam's work, right?

Reply:
No. This won't go away.